File: tagotip-node/src/validate.ts
import {
  MAX_VARNAME_LEN,
  MAX_SERIAL_LEN,
  MAX_GROUP_LEN,
  MAX_META_KEY_LEN,
  MAX_UNIT_LEN,
} from "./consts.ts";
import { TagotipError, type ParseErrorKind } from "./error.ts";

function fail(kind: ParseErrorKind, pos: number): never {
  throw new TagotipError(kind, pos);
}

function isLowercaseAlnumUnderscore(ch: string): boolean {
  return (
    (ch >= "a" && ch <= "z") ||
    (ch >= "0" && ch <= "9") ||
    ch === "_"
  );
}

export function validateVarname(name: string, pos: number): void {
  if (name.length === 0 || name.length > MAX_VARNAME_LEN) {
    fail("invalid_field", pos);
  }
  for (let i = 0; i < name.length; i++) {
    if (!isLowercaseAlnumUnderscore(name[i])) {
      fail("invalid_field", pos);
    }
  }
}

export function validateSerial(serial: string, pos: number): void {
  if (serial.length === 0 || serial.length > MAX_SERIAL_LEN) {
    fail("invalid_serial", pos);
  }
  for (let i = 0; i < serial.length; i++) {
    const ch = serial[i];
    if (
      !(
        (ch >= "a" && ch <= "z") ||
        (ch >= "A" && ch <= "Z") ||
        (ch >= "0" && ch <= "9") ||
        ch === "-" ||
        ch === "_"
      )
    ) {
      fail("invalid_serial", pos);
    }
  }
}

export function validateGroup(group: string, pos: number): void {
  if (group.length === 0 || group.length > MAX_GROUP_LEN) {
    fail("invalid_field", pos);
  }
  for (let i = 0; i < group.length; i++) {
    if (!isLowercaseAlnumUnderscore(group[i])) {
      fail("invalid_field", pos);
    }
  }
}

export function validateMetaKey(key: string, pos: number): void {
  if (key.length === 0 || key.length > MAX_META_KEY_LEN) {
    fail("invalid_metadata", pos);
  }
  for (let i = 0; i < key.length; i++) {
    if (!isLowercaseAlnumUnderscore(key[i])) {
      fail("invalid_metadata", pos);
    }
  }
}

export function validateUnit(unit: string, pos: number): void {
  if (unit.length === 0 || unit.length > MAX_UNIT_LEN) {
    fail("invalid_field", pos);
  }
}

export function validateNumber(s: string, pos: number): void {
  let i = 0;
  if (i < s.length && s[i] === "-") i++;
  if (i >= s.length) fail("invalid_variable", pos);

  if (s[i] === "0") {
    i++;
  } else if (s[i] >= "1" && s[i] <= "9") {
    i++;
    while (i < s.length && s[i] >= "0" && s[i] <= "9") i++;
  } else {
    fail("invalid_variable", pos);
  }

  if (i < s.length && s[i] === ".") {
    i++;
    if (i >= s.length || s[i] < "0" || s[i] > "9") {
      fail("invalid_variable", pos);
    }
    while (i < s.length && s[i] >= "0" && s[i] <= "9") i++;
  }

  if (i !== s.length) fail("invalid_variable", pos);
}

File: tagotip-codec/src/validate.rs
use crate::consts;
use crate::error::{ParseError, ParseErrorKind};

/// Validate a variable name: lowercase a-z, digits, underscore. Max 100 bytes.
pub fn validate_varname(name: &str, pos: usize) -> Result<(), ParseError> {
    if name.is_empty() {
        return Err(ParseError::new(ParseErrorKind::InvalidField, pos));
    }
    if name.len() > consts::MAX_VARNAME_LEN {
        return Err(ParseError::new(ParseErrorKind::InvalidField, pos));
    }
    for &b in name.as_bytes() {
        if !(b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'_') {
            return Err(ParseError::new(ParseErrorKind::InvalidField, pos));
        }
    }
    Ok(())
}

/// Validate a serial number: alphanumeric, hyphen, underscore. Max 100 bytes.
pub fn validate_serial(serial: &str, pos: usize) -> Result<(), ParseError> {
    if serial.is_empty() {
        return Err(ParseError::new(ParseErrorKind::InvalidSerial, pos));
    }
    if serial.len() > consts::MAX_SERIAL_LEN {
        return Err(ParseError::new(ParseErrorKind::InvalidSerial, pos));
    }
    for &b in serial.as_bytes() {
        if !(b.is_ascii_alphanumeric() || b == b'-' || b == b'_') {
            return Err(ParseError::new(ParseErrorKind::InvalidSerial, pos));
        }
    }
    Ok(())
}

/// Validate a group name: same rules as variable name. Max 100 bytes.
pub fn validate_group(group: &str, pos: usize) -> Result<(), ParseError> {
    if group.is_empty() {
        return Err(ParseError::new(ParseErrorKind::InvalidField, pos));
    }
    if group.len() > consts::MAX_GROUP_LEN {
        return Err(ParseError::new(ParseErrorKind::InvalidField, pos));
    }
    for &b in group.as_bytes() {
        if !(b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'_') {
            return Err(ParseError::new(ParseErrorKind::InvalidField, pos));
        }
    }
    Ok(())
}

/// Validate a metadata key: same rules as variable name. Max 100 bytes.
pub fn validate_meta_key(key: &str, pos: usize) -> Result<(), ParseError> {
    if key.is_empty() {
        return Err(ParseError::new(ParseErrorKind::InvalidMetadata, pos));
    }
    if key.len() > consts::MAX_META_KEY_LEN {
        return Err(ParseError::new(ParseErrorKind::InvalidMetadata, pos));
    }
    for &b in key.as_bytes() {
        if !(b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'_') {
            return Err(ParseError::new(ParseErrorKind::InvalidMetadata, pos));
        }
    }
    Ok(())
}

/// Validate a unit string: non-empty, max 25 bytes.
pub fn validate_unit(unit: &str, pos: usize) -> Result<(), ParseError> {
    if unit.is_empty() {
        return Err(ParseError::new(ParseErrorKind::InvalidField, pos));
    }
    if unit.len() > consts::MAX_UNIT_LEN {
        return Err(ParseError::new(ParseErrorKind::InvalidField, pos));
    }
    Ok(())
}

/// Validate number format per spec: `-?(0|[1-9][0-9]*)(\.[0-9]+)?`
pub fn validate_number(s: &str, pos: usize) -> Result<(), ParseError> {
    let bytes = s.as_bytes();
    let mut i = 0;

    if i < bytes.len() && bytes[i] == b'-' {
        i += 1;
    }

    if i >= bytes.len() {
        return Err(ParseError::new(ParseErrorKind::InvalidVariable, pos));
    }

    // int-part: "0" / (%x31-39 *DIGIT)
    if bytes[i] == b'0' {
        i += 1;
    } else if bytes[i] >= b'1' && bytes[i] <= b'9' {
        i += 1;
        while i < bytes.len() && bytes[i].is_ascii_digit() {
            i += 1;
        }
    } else {
        return Err(ParseError::new(ParseErrorKind::InvalidVariable, pos));
    }

    // Optional decimal fraction
    if i < bytes.len() && bytes[i] == b'.' {
        i += 1;
        if i >= bytes.len() || !bytes[i].is_ascii_digit() {
            return Err(ParseError::new(ParseErrorKind::InvalidVariable, pos));
        }
        while i < bytes.len() && bytes[i].is_ascii_digit() {
            i += 1;
        }
    }

    if i != bytes.len() {
        return Err(ParseError::new(ParseErrorKind::InvalidVariable, pos));
    }

    Ok(())
}
